---
title: "Ch. 6: Suckers' Niche JDSM"
author: "Kurt Riggin"
output: html_document
date: "2025-05-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(tidyverse)
library(Hmsc)
library(abind)
load("~/R/catostomidae_JDSM_exploration/Chapter_6_Workspace.RData")
```

## Question: How does temperature and species niches impact the distribution of Suckers across Northeastern United States?


## Methods: 

The same data set taken from Chapter 5 that includes species presence, fixed environmental variables, and site reference condition was used in this analysis. Additionally, trait data taken from Frimpong and Angermeier (2008) was subsetted to include data on temperature tolerance (minimum and maximum temperature), substrate types, and stream current speed for 23 out of 28 species in the original data set mentioned previously. Of those 23, 3 were missing data on temperature tolerance, setting the effective species amount to 20. A phylogeny was also constructed from a mitochondrial genome data set accessed from GenBank following a paper by Yang, Mayden & Naylor (2024), which only included 13 out of 28 species in the original dataset. These 13 species were included in all 3 datasets with matching scientific names, and were thus used in the analysis. 


```{r, echo = F}
plot(tree)
```



Significant data wrangling was performed to match species names between the phylogeny, trait data, and original data set. A joint species distribution model was then built using the Hmsc package following the framework from Chapter 6 of *Joint Species Distribution Modelling with Applications in R* by Ovaskainen and Abrego. Temperature was selected as the environmental covariate and temperature tolerance was also selected as a trait related to this variable. MCMC was ran using the following parameters in the following code, and convergence was examined for both fixed and random effects using plots, effective sample size, and gelman's diganostic. 


```{r, message = F}
####Setup model
Y <- as.matrix(data_C %>% select(carpiodes.carpio:moxostoma.rupiscartes))
XData <- data.frame(temp = data_C$ws_temp)
TrData <- data.frame(temptol = traits$MAXTEMP)

rownames(TrData) <- matchedscinames #rownames of TrData must match species names

studyDesign <- data.frame(site_id = as.factor(data_C$SITE_ID))
rL <- HmscRandomLevel(units = studyDesign$site_id)

XFormula <- ~ temp
TrFormula <- ~ temptol


model <- Hmsc(Y = Y, XData = XData, XFormula = XFormula, TrData = TrData, TrFormula = TrFormula,
              phyloTree = tree, distr = "probit", studyDesign = studyDesign, ranLevels = list(site_id = rL))

nChains = 2
samples = 1000
thin = 50
transient = 5000
verbose = 1000
nParallel = 2
# modelsample <- sampleMcmc(model, thin = thin, samples = samples, transient = transient,
#                           nChains = nChains, verbose = verbose, nParallel = nParallel,
#                           initPar = "fixed effects")

```


Gamma parameters were then explored, to model the influence of temperature on the temperature tolerance to determine if there was a relationship between environmental conditions and species traits.

```{r}
postGamma = getPostEstimate(modelsample, parName="Gamma")
```


Finally the model's predictive power was assessed using the validation set of reference sites by summing the occurrence probabilities across posterior distributions for species and using the mean of the summed distribution as a measure of "expected" values. The expected value was then compared to the observed amount of species by summing probabilities across sites, leading to an observed/expected score for analysis.


```{r, message = FALSE, warning = FALSE}
####Validate model predictions
# setup predictions for validation dataset
studyDesign2 <- data.frame(site_id = as.factor(data_V$SITE_ID))
rL2 <- HmscRandomLevel(units = studyDesign$site_id)
XData.grid <- data.frame(temp = data_V$ws_temp)
Gradient.grid <- prepareGradient(modelsample, XDataNew = XData.grid,
                                 sDataNew = list(site_id = rL))
predYnr <- predict(modelsample, Gradient = Gradient.grid)


# posterior occurrence probability for each site
EpredY <- as.data.frame(apply(abind(predYnr, along = 3), c(1,2), mean))
data_V$observed <- rowSums(data_V[, 17:29])
EpredY$SITE_ID <- data_V$SITE_ID
posteriordata <- data_V %>% select(SITE_ID, ws_temp, observed) %>% left_join(EpredY, by = "SITE_ID")
posteriordata$expected <- rowSums(posteriordata[, 3:15])
posteriordata$oe <- posteriordata$observed/posteriordata$expected
```







## Results:


MCMC convergence diagnostics for the model were not satisfactory, even among changing model parameters to allow for much longer computing time. This was most likely due to the fact that the subsetted data was very sparse among certain species, and thus more difficult to pull a posterior distriubtion sample from. 

```{r, echo = F, message=F}

effectiveSize(modelpost$Beta)
summary(modelpost$Beta)

```

Gamma parameter analysis demonstrated that there was a positive relationship between temperature of the site and temperature tolerance of species, however only at a low support level of 0.55. This indicates that there could be something suggestive about the relationship between temperature and temperature tolerance as expected, however further convergence of a better fit model would be needed to draw conclusions.


```{r, echo = F}
plotGamma(modelsample, post=postGamma, param="Support", supportLevel = 0.55, covNamesNumbers = c(T,F), trNamesNumbers = c(T,F), colorLevels = 3)
```





Finally, although model convergence was not satisfactory, the model gave validation sites a mean observed/expected value of 0.199. A majority of the sites selected did not have any presence of any species used in model building, which caused a skew in the O/E values towards 0.


```{r, echo = F, message=F}

ggplot(posteriordata, aes(y = oe)) +
  geom_boxplot()

hist(posteriordata$oe)
```






## Challenges:

The data wrangling in particular for this model was particularly lengthy. This was mainly due to the fact of having to match species names in the phylogeny, trait data, and community datasets. Even after careful curation of the data, upon a final analysis, there are still species not included due to a single character difference (phylogeny data has White suckers as Catostomus commersonii whereas community dataset has it as Catostomus commersoni). Rerunning this analysis while looking at all possible typos and mismatched data may be worthwhile as the wrangling took longer than the model building itself. A helpful YouTube video about the model capabilities as well as some provided R scripts can be found here, which was loosely used as a guideline for formatting the data (https://www.youtube.com/watch?v=u07eFE3Uqtg&t=1445s&ab_channel=EarthLabCUBoulder). Another clear challenge was the dependency of matched data causing the data used for the analysis to shrink, which impacted what we could get out of the model. A good next step would be to implement a taxonomy as opposed to a phylogeny to see what species would still be left in the analysis and how that could impact our results.






